### PR TITLE
Get Resumes by name

### PR DESCRIPTION
**Pull request by gustavo-rojas-jalafund**
[AB#270](https://dev.azure.com/funjaladev/7867a1c5-5ac7-4c7f-baff-87d740bf4fc1/_workitems/edit/270) https://dev.azure.com/funjaladev/DevLevel3/_workitems/edit/270

### CODE REVIEW

File: src/dal/Jalasoft.TeamUp.Resumes.DAL.Interfaces/IRepository.cs [hunk 1]
```diff
@@ -14,5 +15,7 @@ public interface IRepository<T>
         public void Update(int id, T updateObject);
 
         public void Delete(int id);
+
+        public IEnumerable<Resume> GetByName(string name);
```
Comment (paola-quintanilla-jalafund) [inline src/dal/Jalasoft.TeamUp.Resumes.DAL.Interfaces/IRepository.cs:19]: this line should be in IResumeSQLRepository and IResumesInMemoryRepository

File: src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesInMemoryRepository.cs [hunk 2]
```diff
@@ -113,5 +113,10 @@ public void Update(int id, Resume updateObject)
         {
             throw new NotImplementedException();
         }
+
+        public IEnumerable<Resume> GetByName(string name)
+        {
+            return Resumes;
```
Comment (paola-quintanilla-jalafund) [inline src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesInMemoryRepository.cs:152]: you should return only resumes that match the name

File: src/dal/Jalasoft.TeamUp.Resumes.DAL.Interfaces/IRepository.cs (hunk 1 above)
Comment (Raul-Gamarra-jalafund) [inline src/dal/Jalasoft.TeamUp.Resumes.DAL.Interfaces/IRepository.cs:19]: the return should be a Generic IEnumerable in this interface due that any repo with any entity can use it

Comment (Raul-Gamarra-jalafund) [reply src/dal/Jalasoft.TeamUp.Resumes.DAL.Interfaces/IRepository.cs:19]: Actually this method should be in the ResumesRepository specific domain interface since this is probably not a method any repo would use

File: src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesInMemoryRepository.cs (hunk 2 above)
Comment (Raul-Gamarra-jalafund) [inline src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesInMemoryRepository.cs:152]: this method should be returning only the resumes that match the name, remember is an actual repo, is not a mock

File: tests/api/Jalasoft.TeamUp.Resumes.API.Tests/GetResumesByNameTests.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                            Name = "HTML"
+                        },
+                    }
+                }
+            };
+            return resumes;
+        }
+
+        [Fact]
+        public void GetResumeByName_ValidId_Resume()
```
Comment (Raul-Gamarra-jalafund) [inline tests/api/Jalasoft.TeamUp.Resumes.API.Tests/GetResumesByNameTests.cs:106]: GetResumeByName_ValidId_Resume
This shouldn't be "validID" since we are not receiving an ID we are receiving a name

File: tests/api/Jalasoft.TeamUp.Resumes.API.Tests/GetResumesByNameTests.cs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            // Arrange
+            var request = this.mockHttpContext.Request;
+            this.mockService.Setup(service => service.GetByName("Gustavo")).Returns(GetTestResumes().ToArray());
+
+            // Act
+            var response = this.getResumesByName.Run(request, "Gustavo");
+
+            // Assert
+            var okObjectResult = Assert.IsType<OkObjectResult>(response);
+            Assert.IsType<Resume[]>(okObjectResult.Value);
```
Comment (Raul-Gamarra-jalafund) [inline tests/api/Jalasoft.TeamUp.Resumes.API.Tests/GetResumesByNameTests.cs:117]: please validate the response code in this test as well

File: tests/api/Jalasoft.TeamUp.Resumes.API.Tests/GetResumesByNameTests.cs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        }
+
+        [Fact]
+        public void GetResumesByName_InvalidResume_NotFound()
```
Comment (Raul-Gamarra-jalafund) [inline tests/api/Jalasoft.TeamUp.Resumes.API.Tests/GetResumesByNameTests.cs:121]: this should be InvalidResume as we are sending a Name not a resume

File: tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        [Fact]
+        public void GetResumesByName_ItemsExist_ResumesArray()
+        {
+            // Arrange
+            this.mockRepository.Setup(repository => repository.GetAll()).Returns(GetTestResumes().ToList());
+
+            // Act
+            var result = this.resumesService.GetResumes();
```
Comment (Raul-Gamarra-jalafund) [inline tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs:120]: you should be testing GetByNAme here, not GetResumes

File: tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+            // Assert
+            Assert.Equal(2, result.Length);
+        }
+
+        [Fact]
+        public void GetResume_ValidId_SingleResume()
```
Comment (Raul-Gamarra-jalafund) [inline tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs:127]: we are not receiving an ID the name should not be "validID"

File: tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        {
+            // Arrange
+            var stubResume = new Resume { Id = 1 };
+            this.mockRepository.Setup(repository => repository.GetById(1)).Returns(stubResume);
```
Comment (Raul-Gamarra-jalafund) [inline tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs:131]: we shouln't be using GetById for this mock, we are testing GetResumeByName

File: tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+            // Act
+            var result = this.resumesService.GetResume(1);
```
Comment (Raul-Gamarra-jalafund) [inline tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs:134]: we shouln't be calling GetResumebyId, we are testing GetResumeByName

File: tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs (hunk 7 above)
Comment (Raul-Gamarra-jalafund) [reply tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs:127]: we probably don't need to test a single Resume in this case either since we return a Resume array

File: tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+            // Assert
+            Assert.NotNull(result);
+        }
+
+        [Fact]
+        public void GetResume_UnexistentID_ReturnNull()
```
Comment (Raul-Gamarra-jalafund) [inline tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs:141]: We are not testing GetResume, this name is not appropriate

File: tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+        {
+            Resume resp = null;
+
+            // Arrange
+            this.mockRepository.Setup(repository => repository.GetById(1)).Returns(resp);
```
Comment (Raul-Gamarra-jalafund) [inline tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs:146]: we are not testing GetByID we shouldn't set up this for the mock

File: tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+
+            // Assert
+            Assert.Null(this.resumesService.GetResume(1));
```
Comment (Raul-Gamarra-jalafund) [inline tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs:149]: we are not testing GetResume, we are testing GetResumeByName

File: tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        [Fact]
+        public void GetResumesByName_NotExit_NameResume()
```
Comment (Raul-Gamarra-jalafund) [inline tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs:153]: this name is not related to de implementation please modify to fit the actual code and follow the guidelines the team is using if any doubts reach to @niver-copa-jalafund  or @lilian-delgado-jalafund

File: tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        {
+            // Arrange
+            this.mockRepository.Setup(repository => repository.GetByName("Gustavo")).Throws(new ResumesException(ResumesErrors.NotFound));
+
+            // Assert
+            Assert.Throws<ResumesException>(() => this.resumesService.GetByName("Gustavo"));
```
Comment (Raul-Gamarra-jalafund) [inline tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs:159]: we should be testing this at this layer as the exceptions are only handled in the api layer

File: tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+
+            // Assert
+            Assert.Null(this.resumesService.GetResume(1));
+        }
+
+        [Fact]
+        public void GetResumesByName_NotExit_NameResume()
+        {
+            // Arrange
+            this.mockRepository.Setup(repository => repository.GetByName("Gustavo")).Throws(new ResumesException(ResumesErrors.NotFound));
```
Comment (Raul-Gamarra-jalafund) [inline tests/core/Jalasoft.TeamUp.Resumes.Core.Tests/GetResumesByNameCoreTests.cs:156]: the repo does not return not found exception in any use case

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetResumesByName.cs [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+            this.resumesService = resumesService;
+        }
+
+        [FunctionName("GetResumesByName")]
+        [OpenApiOperation(operationId: "GetResumesByName", tags: new[] { "Resumes" })]
+        [OpenApiParameter(name: "name", In = ParameterLocation.Path, Required = true, Type = typeof(string), Description = "The name of the skill to search by.")]
+        [OpenApiResponseWithBody(statusCode: HttpStatusCode.OK, contentType: "application/json", bodyType: typeof(Resume[]), Description = "Successful response")]
+        [OpenApiResponseWithoutBody(statusCode: HttpStatusCode.NotFound, Description = "Resource not found")]
+        public IActionResult Run(
+        [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "v1/resumes/{name}")] HttpRequest req, string name)
```
Comment (cristian-chavez-jalafund) [inline src/api/Jalasoft.TeamUp.Resumes.API/Controllers/GetResumesByName.cs:30]: should the route be this?
what if there are spaces, does it work? is it following restful best practices?
does the name have to be path param or query param?

File: src/api/Jalasoft.TeamUp.Resumes.API/local.settings.json [hunk 17]
```diff
@@ -3,7 +3,7 @@
   "Values": {
     "AzureWebJobsStorage": "UseDevelopmentStorage=true",
     "FUNCTIONS_WORKER_RUNTIME": "dotnet",
-    "SQLConnetionString": "Server = DESKTOP-L33PS1L; DataBase = TeamUpResumes; User ID = AppConnection; Password = 12345678;"
+    "SQLConnetionString": "Server = DESKTOP-33OMQKP\\SQLEXPRESS; DataBase = TeamUpResumes; User ID = AppConnection; Password = 12345678;"
```
Comment (cristian-chavez-jalafund) [inline src/api/Jalasoft.TeamUp.Resumes.API/local.settings.json]: no need to update this file

**Review by Raul-Gamarra-jalafund — DISMISSED** [review]

**Review by Raul-Gamarra-jalafund — DISMISSED** [review]

**Review by freddy-perez-jalafund — DISMISSED** [review]

**Review by paola-quintanilla-jalafund — DISMISSED** [review]

**Review by Raul-Gamarra-jalafund — DISMISSED** [review]

**Review by freddy-perez-jalafund — DISMISSED** [review]

**Review by paola-quintanilla-jalafund — APPROVED** [review]

**Review by Raul-Gamarra-jalafund — APPROVED** [review]

**Review by freddy-perez-jalafund — APPROVED** [review]

File: src/dal/Jalasoft.TeamUp.Resumes.DAL.Interfaces/IRepository.cs (hunk 1 above)
Comment (gustavo-rojas-jalafund) [reply src/dal/Jalasoft.TeamUp.Resumes.DAL.Interfaces/IRepository.cs:19]: It's right, done.

File: tests/api/Jalasoft.TeamUp.Resumes.API.Tests/GetResumesByNameTests.cs (hunk 3 above)
Comment (gustavo-rojas-jalafund) [reply tests/api/Jalasoft.TeamUp.Resumes.API.Tests/GetResumesByNameTests.cs:106]: It's right, corrected.

File: src/dal/Jalasoft.TeamUp.Resumes.DAL.Interfaces/IRepository.cs (hunk 1 above)
Comment (gustavo-rojas-jalafund) [reply src/dal/Jalasoft.TeamUp.Resumes.DAL.Interfaces/IRepository.cs:19]: Done

File: src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesInMemoryRepository.cs (hunk 2 above)
Comment (gustavo-rojas-jalafund) [reply src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesInMemoryRepository.cs:152]: Done
